src/zenml/materializers/built_in_materializer.py

```diff
@@ -66,11 +71,246 @@ def handle_input(self, data_type: Type[Any]) -> Any:
+            self.metadata_path
+        ):
+            raise RuntimeError(
+                f"Materialization of type {data_type} failed. Expected either"
+                f"{self.data_path} or {self.metadata_path} to exist."
+            )
+
+        # If the data was serialized as JSON, deserialize it.
+        if os.path.exists(self.data_path):
+            outputs = yaml_utils.read_json(self.data_path)
+
+        # Otherwise, use the metadata to reconstruct the data as a list.
+        else:
+            metadata = yaml_utils.read_json(self.metadata_path)
+            outputs = []
+            for path_, type_str in zip(metadata["paths"], metadata["types"]):
+                type_ = find_type_by_str(type_str)
+                materializer_class = default_materializer_registry[type_]
+                mock_artifact = DataArtifact()
+                mock_artifact.uri = path_
+                materializer = materializer_class(mock_artifact)
+                element = materializer.handle_input(type_)
+                outputs.append(element)
+
+        # Cast the data to the correct type.
+        if issubclass(data_type, dict) and not isinstance(outputs, dict):
+            keys, values = outputs
+            return dict(zip(keys, values))
+        if issubclass(data_type, tuple):
+            return tuple(outputs)
+        if issubclass(data_type, set):
+            return set(outputs)
+        return outputs
+
+    def handle_return(self, data: Any) -> None:
+        """Materialize a built-in container object.
+
+        If the object can be serialized to JSON, serialize it.
+
+        Otherwise, use the `default_materializer_registry` to find the correct
+        materializer for each element and materialize each element into a
+        subdirectory.
+
+        Tuples and sets are cast to list before materialization.
+
+        For non-serializable dicts, materialize keys/values as separate lists.
+
+        Args:
+            data: The built-in container object to materialize.
+
+        Raises:
+            Exception: If any exception occurs, it is raised after cleanup.
+        """
+        super().handle_return(data)
+
+        # tuple and set: handle as list.
+        if isinstance(data, tuple) or isinstance(data, set):
+            data = list(data)
+
+        # If the data is serializable, just write it into a single JSON file.
+        if _is_serializable(data):
+            yaml_utils.write_json(self.data_path, data)
+
+        # non-serializable dict: Handle as non-serializable list of lists.
+        if isinstance(data, dict):
+            data = [list(data.keys()), list(data.values())]
+
+        # non-serializable list: Materialize each element into a subfolder.
+        # Get path, type, and corresponding materializer for each element.
+        paths, types, materializers = [], [], []
+        for i, element in enumerate(data):
+            element_path = os.path.join(self.artifact.uri, str(i))
```

Comment:
This might lead to strange results when using a remote artifact store on a windows machine. The artifact uri will be something like `s3://my_bucket/step_1` and will be joined using backslashes: `s3://my_bucket/step_1\\3`.

Reply:
ugh why do we support Windows again? 😆

Reply:
Actually I remembered that I implemented something a while ago to catch this exact case, so I think this should actually be fine here.

Reply:
Okay perfect! 😄